aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,253 @@
+
+    .. automethod:: cancel_registration
+
+    .. automethod:: get_registration_fields
+
+    .. automethod:: register
+
+    """
+
+    @asyncio.coroutine
+    def get_client_info(self, client):
+        """
+        A query is sent to the server o obtain the client's data stored at the
+        server.
+
+        :param client: Client from which the query is sent.
+        :type client: :class:`aioxmpp.Client`
+
+        :return: IQ response :attr:`~.IQ.payload` or :data:`None`
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.GET,
+            payload=xso.Query()
+        )
+
+        return (yield from client.send(iq))
+
+    @asyncio.coroutine
+    def change_pass(self,
+                    client,
+                    new_pass,
+                    aux_fields=None):
+        """
+        Change the client password for 'new_pass'.
+
+        :param client: Client who wants to change the password.
+        :type client: :class:`aioxmpp.Client`
+
+        :param new_pass: New password of the client.
+        :type new_pass: :class:`str`
+
+        :param aux_fields: Auxiliary fields in case additional info is needed.
+        :type aux_fields: :class:`dict`
+
+        :return:`None`
+
+        If the query fails for whatever reason, it will raise an exception
+        from inside client.send(iq).
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.username = client.local_jid.localpart
+        iq.payload.password = new_pass
+
+        if aux_fields is not None:
+            for key, value in aux_fields.items():
+                setattr(iq.payload, key, value)
+        yield from client.send(iq)
+
+    @asyncio.coroutine
+    def cancel_registration(self, client):
+        """
+        Cancels the currents client's account with the server.
+
+        :param client: Client who wants to cancel.
+        :type client: :class:`aioxmpp.Client`
+
+        :return:`None`
+
+        Even if the cancelation is succesful, this method will raise an
+        exception due to he account no longer exists for the server, so the
+        client will fail.
+        To continue with the execution, this method should be surrounded by a
+        try/except statement.
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.remove = True
+        yield from client.send(iq)
+
+    @staticmethod
+    @asyncio.coroutine
+    def connect_and_send(iq, loop, metadata, peer_jid, timeout):
+        options = list((
+            yield from aioxmpp.discover_connectors(
+                peer_jid.domain, loop=loop
+            )
+        ))
+        for host, port, conn in options:
```

Comment:
The loop as written will attempt to send the IQ to each of the options and return the reply of the last option. This is not correct behaviour.